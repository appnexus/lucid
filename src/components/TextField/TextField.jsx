import React from 'react';
import PropTypes from 'react-peek/prop-types';
import _ from 'lodash';
import { lucidClassNames } from '../../util/style-helpers';
import { createClass, omitProps } from '../../util/component-types';
import reducers from './TextField.reducers';
import * as KEYCODE from '../../constants/key-code';

const cx = lucidClassNames.bind('&-TextField');

const { bool, string, func, number, object, oneOfType } = PropTypes;

const TextField = createClass({
	displayName: 'TextField',

	statics: {
		peek: {
			description: `
				TextField should cover all your text input needs. It is able to handle
				single and multi line inputs.
			`,
			categories: ['controls', 'text'],
		},
	},

	reducers,

	propTypes: {
		style: object`
			Styles that are passed through to native control.
		`,

		isMultiLine: bool`
			Set the TextField to multi line mode. Under the hood this will use a
			\`textarea\` instead of an \`input\` if set to \`true\`.
		`,

		isDisabled: bool`
			Disables the TextField by greying it out.
		`,

		rows: number`
			Initial number of rows a multi line TextField should have. Ignored when
			not in multi-line mode.
		`,

		className: string`
			Class names that are appended to the defaults.
		`,

		onChange: func`
			Fires an event every time the user types text into the TextField.
			Signature: \`(value, { event, props }) => {}\`
		`,

		onBlur: func`
			Fires an on the \`input\`'s onBlur.  Signature:
			\`(currentValue, { event, props }) => {}\`
		`,

		onChangeDebounced: func`
			Fires an event, debounced by \`debounceLevel\`, when the user types text
			into the TextField.  Signature: \`(value, { event, props }) => {}\`
		`,

		onKeyDown: func`
			Fires an event on every keydown Signature: \`({ event, props }) => {}\`
		`,

		onSubmit: func`
			Fires an event when the user hits "enter" from the TextField. You
			shouldn't use it if you're using \`isMultiLine\`.  Signature:
			\`(value, { event, props }) => {}\`
		`,

		value: oneOfType([number, string])`
			Set the value of the input.
		`,

		debounceLevel: number`
			Number of milliseconds to debounce the \`onChangeDebounced\` callback.
			Only useful if you provide an \`onChangeDebounced\` handler.
		`,

		lazyLevel: number`
			Set the holding time, in milliseconds, that the component will wait if
			the user is typing and the component gets a new \`value\` prop.  Any time
			the user hits a key, it starts a timer that prevents state changes from
			flowing in to the component until the timer has elapsed.  This was
			heavily inspired by the
			[lazy-input](https:/docs.npmjs.com/package/lazy-input) component.
		`,
	},

	getDefaultProps() {
		return {
			style: null,
			isDisabled: false,
			isMultiLine: false,
			onBlur: _.noop,
			onChange: _.noop,
			onChangeDebounced: _.noop,
			onSubmit: _.noop,
			rows: 5,
			debounceLevel: 500,
			lazyLevel: 1000,
			value: '',
		};
	},

	getInitialState() {
		return {
			value: this.props.value,
		};
	},

	componentWillMount() {
		// Because we want the debounceLevel to be configurable, we can't put the
		// debounced handler directly on the react class, so we set it up right
		// before mount
		this._isMounted = true;
		this._handleChangeDebounced = _.debounce((...args) => {
			this.props.onChangeDebounced(...args);
		}, this.props.debounceLevel);

		this._releaseHold = _.debounce(() => {
			if (!this._isMounted) {
				return;
			}
			this.setState({ isHolding: false });
		}, this.props.lazyLevel);

		this._updateWhenReady = _.debounce(newValue => {
			if (!this._isMounted) {
				return;
			}
			if (this.state.isHolding) {
				this._updateWhenReady(newValue);
			} else if (newValue !== this.state.value) {
				this.setState({ value: newValue });
			}
		}, this.props.lazyLevel);
	},

	componentWillUnmount() {
		this._isMounted = false;
	},

	componentWillReceiveProps(nextProps) {
		// Allow consumer to optionally control state
		if (_.has(nextProps, 'value')) {
			if (this.state.isHolding) {
				this._updateWhenReady(nextProps.value);
			} else {
				this.setState({ value: nextProps.value });
			}
		}
	},

	handleChange(event) {
		const { onChange, onChangeDebounced } = this.props;

		const value = _.get(event, 'target.value', '');

		this.setState({ value, isHolding: true });
		this._releaseHold();

		onChange(value, { event, props: this.props });

		// Also call the debounced handler in case the user wants debounced change
		// events.
		if (onChangeDebounced !== _.noop) {
			event.persist(); // https://facebook.github.io/react/docs/events.html#event-pooling
			this._handleChangeDebounced(value, { event, props: this.props });
		}
	},

	handleBlur(event) {
		const { onBlur, onChangeDebounced } = this.props;

		const value = _.get(event, 'target.value', '');

		if (onChangeDebounced !== _.noop) {
			this._handleChangeDebounced.flush();
		}
		onBlur(value, { event, props: this.props });
	},

	handleKeyDown(event) {
		const { props, props: { onSubmit, onKeyDown, onChangeDebounced } } = this;
		const value = _.get(event, 'target.value', '');

		// If the consumer passed an onKeyDown, we call it
		if (onKeyDown) {
			onKeyDown({ event, props });
		}

		if (event.keyCode === KEYCODE.Enter) {
			if (onChangeDebounced !== _.noop) {
				this._handleChangeDebounced.flush();
			}

			onSubmit(value, { event, props: this.props });
		}
	},

	focus() {
		/* istanbul ignore next */
		this.refs.nativeElement.focus();
	},

	render() {
		const {
			className,
			isDisabled,
			isMultiLine,
			rows,
			style,
			...passThroughs
		} = this.props;

		const { value } = this.state;

		const finalProps = {
			...omitProps(passThroughs, TextField, ['children']),
			className: cx(
				'&',
				{
					'&-is-disabled': isDisabled,
					'&-is-multi-line': isMultiLine,
					'&-is-single-line': !isMultiLine,
				},
				className
			),
			disabled: isDisabled,
			onChange: this.handleChange,
			onBlur: this.handleBlur,
			onKeyDown: this.handleKeyDown,
			style,
			rows,
			value,
			ref: ref => (this.refs = { nativeElement: ref }),
		};

		return isMultiLine ? (
			<textarea {...finalProps} />
		) : (
			<input type="text" {...finalProps} />
		);
	},
});

export default TextField;
