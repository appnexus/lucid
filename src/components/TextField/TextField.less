@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

@TextField-size-padding: @size-S;
@TextField-size-height: @size-standard-height;
@TextField-color-font: #333;

.lucid-TextField {
	.box-sizing();
	.inset-box-shadow();

	background-color: @color-white;
	border-radius: @size-borderRadius;
	border: @border-standardBorder;
	color: @TextField-color-font;
	font-family: @fontFamily;
	font-size: @fontSize;
	margin: 0;
	padding: 0 @TextField-size-padding;
	width: 100%;

	&:focus {
		outline: none;
		.focus-glow();
		border-color: @color-primary;
	}

	&&-is-single-line {
		vertical-align: middle;
		height: @TextField-size-height;
	}

	&&-is-multi-line {
		vertical-align: top;
		resize: both;
		padding-top: @TextField-size-padding - 2px; // Handle padding appropriately for `textarea`s
	}

	&&-is-disabled {
		cursor: not-allowed;
		background-color: @color-gray-5;
		border-color: @color-gray-10;
		color: @color-gray-25;
	}
}
