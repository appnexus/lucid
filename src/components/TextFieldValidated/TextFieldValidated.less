@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

.lucid-TextFieldValidated {
	width: 100%;

	&.lucid-Validation-is-error {
		.lucid-TextField {
			.inset-box-shadow(@featured-color-danger);
			border-color: @featured-color-danger;
			&:focus {
				.focus-glow(@featured-color-danger-30);
			}
		}
	}
}
