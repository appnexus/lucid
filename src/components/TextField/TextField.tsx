import React from 'react';
import PropTypes from 'react-peek/prop-types';
import _ from 'lodash';
import { lucidClassNames } from '../../util/style-helpers';
import {
	omitProps,
	StandardProps,
	Overwrite,
} from '../../util/component-types';
import reducers from './TextField.reducers';
import * as KEYCODE from '../../constants/key-code';
import { ITextFieldState } from './TextField.reducers';

const cx = lucidClassNames.bind('&-TextField');

const { bool, string, func, number, object, oneOfType } = PropTypes;

export interface ITextFieldProps extends StandardProps {
	/** Set the TextField to multi line mode. Under the hood this will use a `textarea` instead of an `input` if set to `true`. */
	isMultiLine?: boolean;

	/** Disables the TextField by greying it out. */
	isDisabled?: boolean;

	/** Initial number of rows a multi line TextField should have. Ignored when
		not in multi-line mode. */
	rows?: number;

	/** Fires an event every time the user types text into the TextField. */
	onChange?: (
		value: string,
		{
			event,
			props,
		}: {
			event: React.FormEvent;
			props: ITextFieldProps;
		}
	) => void;

	/** Fires an on the `input`'s onBlur. */
	onBlur?: (
		currentValue: string,
		{
			event,
			props,
		}: {
			event: React.FocusEvent;
			props: ITextFieldProps;
		}
	) => void;

	/** Fires an event, debounced by `debounceLevel` when the user types text
		into the TextField. */
	onChangeDebounced?: (
		value: string,
		{
			event,
			props,
		}: {
			event: React.FormEvent;
			props: ITextFieldProps;
		}
	) => void;

	/** Fires an event on every keydown */
	onKeyDown?: ({
		event,
		props,
	}: {
		event: React.KeyboardEvent;
		props: ITextFieldProps;
	}) => void;

	/** Fires an event when the user hits "enter" from the TextField. You shouldn't use it if you're using `isMultiLine`. */
	onSubmit?: (
		value: string,
		{
			event,
			props,
		}: {
			event: React.FormEvent;
			props: ITextFieldProps;
		}
	) => void;

	/** Set the value of the input. */
	value: string | number;

	/** Number of milliseconds to debounce the `onChangeDebounced` callback. Only useful if you provide an `onChangeDebounced` handler. */
	debounceLevel?: number;

	/** Set the holding time, in milliseconds, that the component will wait if
		the user is typing and the component gets a new `value` prop.  Any time
		the user hits a key, it starts a timer that prevents state changes from
		flowing in to the component until the timer has elapsed.  This was
		heavily inspired by the [lazy-input](https:/docs.npmjs.com/package/lazy-input) component. */
	lazyLevel?: number;
}

export type ITextFieldPropsWithPassThroughs = Overwrite<
	React.InputHTMLAttributes<HTMLInputElement>,
	ITextFieldProps
>;

class TextField extends React.Component<
	ITextFieldPropsWithPassThroughs,
	ITextFieldState,
	{}
> {
	static displayName = 'TextField';
	static peek = {
		description: `
			TextField should cover all your text input needs. It is able to handle
			single and multi line inputs.
		`,
		categories: ['controls', 'text'],
	};
	static propTypes = {
		style: object`
			Styles that are passed through to native control.
		`,

		isMultiLine: bool`
			Set the TextField to multi line mode. Under the hood this will use a
			\`textarea\` instead of an \`input\` if set to \`true\`.
		`,

		isDisabled: bool`
			Disables the TextField by greying it out.
		`,

		rows: number`
			Initial number of rows a multi line TextField should have. Ignored when
			not in multi-line mode.
		`,

		className: string`
			Class names that are appended to the defaults.
		`,

		onChange: func`
			Fires an event every time the user types text into the TextField.
			Signature: \`(value, { event, props }) => {}\`
		`,

		onBlur: func`
			Fires an on the \`input\`'s onBlur.  Signature:
			\`(currentValue, { event, props }) => {}\`
		`,

		onChangeDebounced: func`
			Fires an event, debounced by \`debounceLevel\`, when the user types text
			into the TextField.  Signature: \`(value, { event, props }) => {}\`
		`,

		onKeyDown: func`
			Fires an event on every keydown Signature: \`({ event, props }) => {}\`
		`,

		onSubmit: func`
			Fires an event when the user hits "enter" from the TextField. You
			shouldn't use it if you're using \`isMultiLine\`.  Signature:
			\`(value, { event, props }) => {}\`
		`,

		value: oneOfType([number, string])`
			Set the value of the input.
		`,

		debounceLevel: number`
			Number of milliseconds to debounce the \`onChangeDebounced\` callback.
			Only useful if you provide an \`onChangeDebounced\` handler.
		`,

		lazyLevel: number`
			Set the holding time, in milliseconds, that the component will wait if
			the user is typing and the component gets a new \`value\` prop.  Any time
			the user hits a key, it starts a timer that prevents state changes from
			flowing in to the component until the timer has elapsed.  This was
			heavily inspired by the
			[lazy-input](https:/docs.npmjs.com/package/lazy-input) component.
		`,
	};

	state = {
		value: this.props.value,
		isHolding: false,
		isMounted: false,
	};

	static defaultProps = {
		style: undefined,
		isDisabled: false,
		isMultiLine: false,
		onBlur: _.noop,
		onChange: _.noop,
		onChangeDebounced: _.noop,
		onSubmit: _.noop,
		onKeyDown: _.noop,
		rows: 5,
		debounceLevel: 500,
		lazyLevel: 1000,
		value: '',
	};

	static reducers = reducers;

	private textareaElement = React.createRef<HTMLTextAreaElement>();
	private inputElement = React.createRef<HTMLInputElement>();
	private nativeElement = this.props.isMultiLine
		? this.textareaElement
		: this.inputElement;

	private handleChangeDebounced = _.debounce(
		(
			value: string,
			{
				event,
				props,
			}: {
				event: React.FormEvent;
				props: ITextFieldProps;
			}
		): void => {
			this.props.onChangeDebounced && this.props.onChangeDebounced(value, { event, props });
		},
		this.props.debounceLevel
	);

	private releaseHold = _.debounce((): void => {
		if (!this.state.isMounted) {
			return;
		}
		this.setState({ isHolding: false });
	}, this.props.lazyLevel);

	private updateWhenReady = _.debounce((newValue): void => {
		if (!this.state.isMounted) {
			return;
		}
		if (this.state.isHolding) {
			this.updateWhenReady(newValue);
		} else if (newValue !== this.state.value) {
			this.setState({ value: newValue });
		}
	}, this.props.lazyLevel);

	handleChange = (event: React.FormEvent): void => {
		const { onChange, onChangeDebounced } = this.props;

		const value = _.get(event, 'target.value', '');

		this.setState({ value, isHolding: true });
		this.releaseHold();

		onChange && onChange(value, { event, props: this.props });

		// Also call the debounced handler in case the user wants debounced change
		// events.
		if (onChangeDebounced !== _.noop) {
			event.persist(); // https://facebook.github.io/react/docs/events.html#event-pooling
			this.handleChangeDebounced(value, { event, props: this.props });
		}
	};

	handleBlur = (event: React.FocusEvent): void => {
		const { onBlur, onChangeDebounced } = this.props;

		const value = _.get(event, 'target.value', '');

		if (onChangeDebounced !== _.noop) {
			this.handleChangeDebounced.flush();
		}
		onBlur && onBlur(value, { event, props: this.props });
	};

	handleKeyDown = (event: React.KeyboardEvent): void => {
		const {
			props,
			props: { onSubmit, onKeyDown, onChangeDebounced },
		} = this;
		const value = _.get(event, 'target.value', '');

		// If the consumer passed an onKeyDown, we call it
		if (onKeyDown) {
			onKeyDown({ event, props });
		}

		if (event.keyCode === KEYCODE.Enter) {
			if (onChangeDebounced !== _.noop) {
				this.handleChangeDebounced.flush();
			}

			onSubmit && onSubmit(value, { event, props: this.props });
		}
	};

	focus = (): void => {
		/* istanbul ignore next */
		(this.nativeElement.current as HTMLElement).focus();
	};

	UNSAFE_componentWillMount(): void {
		this.setState({ isMounted: true });
	}

	componentWillUnmount(): void {
		this.setState({ isMounted: false });
	}

	UNSAFE_componentWillReceiveProps(nextProps: ITextFieldProps): void {
		// Allow consumer to optionally control state
		if (_.has(nextProps, 'value')) {
			if (this.state.isHolding) {
				this.updateWhenReady(nextProps.value);
			} else {
				this.setState({ value: nextProps.value });
			}
		}
	}

	render(): React.ReactNode {
		const {
			className,
			isDisabled,
			isMultiLine,
			rows,
			style,
			...passThroughs
		} = this.props;

		const { value } = this.state;

		const finalProps = {
			...omitProps(passThroughs, undefined, [
				...Object.keys(TextField.propTypes),
				'children',
			]),
			className: cx(
				'&',
				{
					'&-is-disabled': isDisabled,
					'&-is-multi-line': isMultiLine,
					'&-is-single-line': !isMultiLine,
				},
				className
			),
			disabled: isDisabled,
			onChange: this.handleChange,
			onBlur: this.handleBlur,
			onKeyDown: this.handleKeyDown,
			style,
			rows,
			value,
		};

		return isMultiLine ? (
			<textarea ref={this.textareaElement} {...finalProps} />
		) : (
			<input type='text' ref={this.inputElement} {...finalProps} />
		);
	}
}

export default TextField;
