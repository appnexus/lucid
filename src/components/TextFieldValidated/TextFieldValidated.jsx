import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import Validation from '../Validation/Validation';
import TextField from '../TextField/TextField';
import reducers from '../TextField/TextField.reducers';
import { lucidClassNames } from '../../util/style-helpers';
import {
	createClass as createReactClass,
	findTypes,
	omitProps,
} from '../../util/component-types';

const cx = lucidClassNames.bind('&-TextFieldValidated');

const { any, object, string } = PropTypes;

/**
 *
 * {"categories": ["controls", "text"], "extend": "TextField", "madeFrom": ["TextField", "Validation"]}
 *
 * A composition of TextField and Validation.
 */
const TextFieldValidated = createReactClass({
	displayName: 'TextFieldValidated',

	reducers,

	components: {
		/**
		 * Content that will be displayed as an error message.
		 */
		Error: createReactClass({
			displayName: 'TextFieldValidated.Error',
			propName: 'Error',
		}),
	},

	propTypes: {
		/**
		 * Passed to the root container.
		 */
		style: object,

		/**
		 * Passed to the root container.
		 */
		className: string,

		/**
		 * Prop alternative to Error child component
		 */
		Error: any,
	},

	getDefaultProps() {
		return TextField.getDefaultProps();
	},

	focus() {
		this.refs.TextField.focus();
	},

	render() {
		const { className, style, ...passThroughs } = this.props;

		const errorChildProps = _.map(
			findTypes(this.props, TextFieldValidated.Error),
			'props'
		);

		return (
			<Validation
				className={cx('&', className)}
				style={style}
				Error={errorChildProps}
			>
				<TextField
					{...omitProps(passThroughs, TextFieldValidated, [], false)}
					ref={ref => this.refs = { TextField: ref }}
				/>
			</Validation>
		);
	},
});

export default TextFieldValidated;
