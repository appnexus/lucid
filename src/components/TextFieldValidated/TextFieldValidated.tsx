import _ from 'lodash';
import React from 'react';
import PropTypes, { shape } from 'prop-types';

import Validation, { IValidationProps } from '../Validation/Validation';
import TextField, {
	ITextFieldPropsWithPassThroughs,
} from '../TextField/TextField';
import reducers from '../TextField/TextField.reducers';
import { lucidClassNames } from '../../util/style-helpers';
import { findTypes } from '../../util/component-types';

const cx = lucidClassNames.bind('&-TextFieldValidated');

const { any, object, string, bool } = PropTypes;

export interface ITextFieldValidatedErrorProps extends IValidationProps {
	description?: string;
}

const TextFieldValidatedError = (_props: ITextFieldValidatedErrorProps): null =>
	null;
TextFieldValidatedError.displayName = 'TextFieldValidated.Error';
TextFieldValidatedError.peek = {
	description: `Content that will be displayed as an error message.`,
};
TextFieldValidatedError.propName = 'Error';

/** TODO: Remove the nonPassThroughs when the component is converted to a functional component */
const nonPassThroughs = [
	'style',
	'className',
	'Error',
	'Info',
	'Success',
	'initialState',
];

export interface ITextFieldValidatedProps
	extends ITextFieldPropsWithPassThroughs {
	Error?: React.ReactNode;
	Info?: string;
	Success?: {
		message: string;
		disappearing?: boolean;
	};
}

export interface ITextFieldValidatedState {
	value: string | number;
}

class TextFieldValidated extends React.Component<
	ITextFieldValidatedProps,
	ITextFieldValidatedState
> {
	static displayName = 'TextFieldValidated';

	static Error = TextFieldValidatedError;

	static peek = {
		description: `A composite of \`TextField\` and \`Validation\`.`,
		categories: ['controls', 'text'],
	};

	static reducers = reducers;

	static propTypes = {
		/**
			Passed to the root container.
		*/
		style: object,

		/**
			Passed to the root container.
		*/
		className: string,

		/**
			Prop alternative to Error child component
		*/
		Error: any,

		/**
			Optional information text that is styled less aggressively than an error
		 */
		Info: string,

		/**
			Optional information text that is styled as "success". Also contains
		  `disappearing` prop for message that appears and fades away.
		 */
		Success: shape({
			message: string,
			disappearing: bool,
		}),
	};

	static defaultProps = TextField.defaultProps;

	private textFieldRef = React.createRef<TextField>();

	focus = () => {
		this.textFieldRef.current && this.textFieldRef.current.focus();
	};

	render() {
		const { className, style, ...passThroughs } = this.props;

		let childProps;

		if (this.props.Error) {
			childProps = _.map(
				findTypes(this.props, TextFieldValidated.Error),
				'props'
			);
		} else if (this.props.Info) {
			childProps = [this.props.Info];
		} else if (this.props.Success) {
			childProps = [this.props.Success?.message];
		}

		const isSuccess =
			!this.props.Error && !this.props.Info && this.props.Success;

		return (
			<Validation
				className={cx('&', className, {
					'-info': !this.props.Error && this.props.Info,
					'-success':
						!this.props.Error && !this.props.Info && this.props.Success,
					'-disappearing': isSuccess && this.props.Success?.disappearing,
				})}
				style={style}
				Error={childProps}
			>
				<TextField
					{...(_.omit(passThroughs, nonPassThroughs) as any)}
					ref={this.textFieldRef}
				/>
			</Validation>
		);
	}
}

export default TextFieldValidated;
