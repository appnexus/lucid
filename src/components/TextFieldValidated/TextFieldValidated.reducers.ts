import { ITextFieldValidatedState } from './TextFieldValidated';

export function onChange(
	state: ITextFieldValidatedState = { value: '' },
	value: string
): ITextFieldValidatedState {
	return {
		...state,
		value: value,
	};
}

export default { onChange };
