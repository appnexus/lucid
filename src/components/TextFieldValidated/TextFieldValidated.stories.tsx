import React, { useState } from 'react';
import { Story, Meta } from '@storybook/react';
import createClass from 'create-react-class';

import TextFieldValidated, {
	ITextFieldValidatedProps,
} from './TextFieldValidated';

export default {
	title: 'Controls/TextFieldValidated',
	component: TextFieldValidated,
	parameters: {
		docs: {
			description: {
				component: TextFieldValidated.peek.description,
			},
		},
	},
} as Meta;

const style = {
	marginBottom: '10px',
};

export const Basic: Story<ITextFieldValidatedProps> = (args) => {
	return <TextFieldValidated {...args} Error='Nope, not even close!' />;
};

export const Debounced: Story<ITextFieldValidatedProps> = (args) => {
	const [value, setValue] = useState('');

	return (
		<div>
			<TextFieldValidated
				{...args}
				style={style}
				value={value}
				onChangeDebounced={(value) => setValue(value)}
				Error={value === 'foo' ? null : 'Please enter "foo"'}
			/>
		</div>
	);
};

export const ErrorTypes: Story<ITextFieldValidatedProps> = (args) => {
	const [value, setValue] = useState('');

	return (
		<div>
			<TextFieldValidated
				{...args}
				style={style}
				value={value}
				onChangeDebounced={() => {}}
				Error={'This is an error'}
			/>
			<TextFieldValidated
				{...args}
				style={style}
				value={value}
				onChangeDebounced={() => {}}
				Error={null}
				Info={'This is an info'}
			/>
		</div>
	);
};
