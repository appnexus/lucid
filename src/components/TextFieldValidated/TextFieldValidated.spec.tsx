import React from 'react';
import { common } from '../../util/generic-tests';
import TextFieldValidated from './TextFieldValidated';
import TextField from '../TextField/TextField';
import { shallow } from 'enzyme';
import assert from 'assert';

describe('TextFieldValidated', () => {
	common(TextFieldValidated);

	describe('props', () => {
		it('should pass through props to TextField', () => {
			const onBlur = () => {};
			const wrapper = shallow(<TextFieldValidated onBlur={onBlur} />);
			assert.equal(wrapper.find(TextField).prop('onBlur'), onBlur);
		});
	});
});
