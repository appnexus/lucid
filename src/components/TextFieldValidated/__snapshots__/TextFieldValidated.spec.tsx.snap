// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextFieldValidated [common] example testing should match snapshot(s) for basic 1`] = `
<Validation
  Error={
    Array [
      Object {
        "children": "Nope, not even close!",
      },
    ]
  }
  className="lucid-TextFieldValidated"
>
  <TextField
    debounceLevel={500}
    isDisabled={false}
    isMultiLine={false}
    lazyLevel={1000}
    onBlur={[Function]}
    onChange={[Function]}
    onChangeDebounced={[Function]}
    onKeyDown={[Function]}
    onSubmit={[Function]}
    rows={5}
    value=""
  />
</Validation>
`;

exports[`TextFieldValidated [common] example testing should match snapshot(s) for debounced 1`] = `
<Validation
  Error={
    Array [
      Object {
        "children": "Please enter \\"foo\\"",
      },
    ]
  }
  className="lucid-TextFieldValidated"
>
  <TextField
    debounceLevel={500}
    isDisabled={false}
    isMultiLine={false}
    lazyLevel={1000}
    onBlur={[Function]}
    onChange={[Function]}
    onChangeDebounced={[Function]}
    onKeyDown={[Function]}
    onSubmit={[Function]}
    rows={5}
    value=""
  />
</Validation>
`;
